flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/processor/ForwardHashExchangeProcessor.java

```diff
@@ -0,0 +1,171 @@
+ * adaptive scheduler is enabled: FORWARD partitioner if the parallelism is the same, or HASH
+ * partitioner if the parallelism is different.
+ *
+ * <p>Its works only for batch job when dynamic-graph is enabled.
+ */
+public class ForwardHashExchangeProcessor implements ExecNodeGraphProcessor {
+
+    @Override
+    public ExecNodeGraph process(ExecNodeGraph execGraph, ProcessorContext context) {
+        if (execGraph.getRootNodes().get(0) instanceof StreamExecNode) {
+            throw new TableException("StreamExecNode is not supported yet");
+        }
+        if (!context.getPlanner().getExecEnv().getConfig().isDynamicGraph()) {
+            return execGraph;
+        }
+        ExecNodeVisitor visitor =
+                new AbstractExecNodeExactlyOnceVisitor() {
+                    @Override
+                    protected void visitNode(ExecNode<?> node) {
+                        visitInputs(node);
+                        if (node instanceof CommonExecExchange) {
+                            return;
+                        }
+                        boolean changed = false;
+                        List<ExecEdge> newEdges = new ArrayList<>(node.getInputEdges());
+                        for (int i = 0; i < node.getInputProperties().size(); ++i) {
+                            InputProperty inputProperty = node.getInputProperties().get(i);
+                            InputProperty.RequiredDistribution requiredDistribution =
+                                    inputProperty.getRequiredDistribution();
+                            if (requiredDistribution.getType()
+                                    != InputProperty.DistributionType.HASH) {
+                                continue;
+                            }
+                            ExecEdge edge = node.getInputEdges().get(i);
+                            if (!hasExchangeInput(edge)) {
+                                InputProperty newInputProperty =
+                                        InputProperty.builder()
+                                                .requiredDistribution(
+                                                        InputProperty.keepInputAsIsDistribution(
+                                                                requiredDistribution))
+                                                .damBehavior(inputProperty.getDamBehavior())
+                                                .priority(inputProperty.getPriority())
+                                                .build();
+                                BatchExecExchange newExchange =
+                                        new BatchExecExchange(
+                                                newInputProperty,
+                                                (RowType) edge.getOutputType(),
+                                                newInputProperty.toString());
+
+                                ExecEdge newEdge1 =
```

Comment:
a suggestion: how about add an updateInput(ExecNodeBase newInput, int inputIndex) to ExecNodeBase, and BatchExecMultiInput rewrite the method, and update the input of the original node. 
we could make the magic of MultiInput internal in BatchMultiInput and make it easier to understand the logic here.

Reply:
It's better we keep the logic in one class